keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,858 @@
+    // If this field is not specified, then the supported number of volumes on this no
+    // +optional
+    Count *int32
+}
+```
+
+#### CSIDriver
+
+A new field, `AllocatableUpdateInterval`, will be added to the `CSIDriverSpec` struct. This field allows a CSI driver to
+specify the interval at which the Kubelet should periodically query a driver's `NodeGetInfo` RPC endpoint to update the
+`CSINode` object. If this field is not set, updates will only occur in response to volume attachment failures as a result of no
+capacity.
+
+```golang
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ...
+    // AllocatableUpdateInterval specifies the interval between periodic updates of
+    // the CSINode allocatable capacity for this driver. If not set, periodic updates
+    // are disabled, and updates occur only upon detecting capacity-related failures.
+    // +optional
+    AllocatableUpdateInterval *metav1.Duration
+}
+```
+
+#### Validation Changes
+
+The `ValidateCSINodeUpdate` function in the API validation code path will be modified to allow updates to the `Allocatable.Count`
+field when the feature gate is enabled:
+
+```golang
+func ValidateCSINodeUpdate(new, old *storage.CSINode) field.ErrorList {
+    allErrs := ValidateCSINode(new)
+    
+    if utilfeature.DefaultFeatureGate.Enabled(features.MutableCSINode) {
+        for _, oldDriver := range old.Spec.Drivers {
+           for _, newDriver := range new.Spec.Drivers {
+                // Allow Allocatable.Count to be modified
+                // Ensure all other fields are unchanged
+            }
+        }
+    } else {
+        // Existing validation logic for when feature gate is disabled
+    }
+    return allErrs
+}
+```
+
+This updated logic allows the `Allocatable.Count` field to be modified when the feature gate is enabled, while ensuring all
+other fields remain immutable. When the feature gate is disabled, it falls back to the existing validation logic for backward
+compatibility.
+
+#### Volume Plugin Manager
+
+A new goroutine will be started in VolumePluginMgr’s `Run()` func if the `AllocatableUpdateInterval` is set to a nonzero value. This goroutine will periodically trigger updates to the `CSINode` object based on the specified interval:
+
+```golang
+func (pm *VolumePluginMgr) Run(stopCh <-chan struct{}) {
+    if pm.csiNodeUpdateInterval > 0 {
+        go wait.Until(pm.updateCSINodeInfo, pm.csiNodeUpdateInterval, stopCh)
+    }
+}
+```
+
+#### NodeInfoManager Interface Extension
+
+The existing `NodeInfoManager` interface will be extended to include a new method for updating the `CSINode` object:
+
+```golang
+// Interface implements an interface for managing labels of a node
+type Interface interface {
+    CreateCSINode() (*storagev1.CSINode, error)
+    ...
+    // UpdateCSINode updates the CSINode object
+    UpdateCSINode() error
+}
+```
+
+#### CSINode Update Mechanism
+
```

Comment:
If the feature flag is not set and `AllocatableUpdateInterval` is set then what's the behavior?

Reply:
When the feature gate is disabled, `AllocatableUpdateInterval` is ignored and `Allocatable.Count` remains static + immutable. Included a table in latest revision for clarity.